### ISSUE TITLE
Multiscale: An FE domain solve for every integration point

### ISSUE THREAD

**Issue by karthichockalingam** [mfem/mfem#4406]
Hello,

We ( @nmnobre ) are working on a multiscale problem, which involves solving the problem at two scales namely macro (coarse) and micro (fine) scale. We want to resolve the fine-scale problem at very macro integration/quadrature point. I am using a modified macro quadrature rule that involves using the stiffness obtained from solving the micro problem located at that quadrature point. I've attached a document [Multiscale.pdf](https://github.com/user-attachments/files/16254245/Multiscale.pdf) that shows the schematic of the method for a 2D Quad first-order Lagrange element. 

There are three main tasks for numerical implementation:

1) Center the micro-domain at every quadrature point in the macro finite element discretization.

2) Project/transfer the macro test functions and apply them as Dirichlet boundary conditions for the micro problem. There are solution fields from the microdomain for each of the macro basis functions. 

3) Transfer the solution field from the micro solve to be used in the macro quadrature rule.

How can I get started? Thank you for your input.

Kind regards,
Karthik.

**Comment by vladotomov** [comment]
Hi @karthichockalingam,

Yes this should be possible with the existing MFEM functionality.

You can start by writing the two solvers and examining the transfer functionality examples in `/miniapps/gslib`.

Let us know if you have a specific question.

**Comment by karthichockalingam** [comment]
@vladotomov thank you for your response. Here [macro_micro_problem.pdf](https://github.com/user-attachments/files/16319884/macro_micro_problem.pdf) both the macro and micro are diffusion problems.  I am planning to create a MultiscaleIntegrator (which is essentially a copy of the DiffusionIntegrator) and call the micro problem at this line

https://github.com/mfem/mfem/blob/0952809e5e330439bd8c320a0e80ae7cae228460/fem/bilininteg.cpp#L895

so it is solved at every integration point of the macro problem. I will look into ```/miniapps/gslib``` for the transfer functionality.

**Comment by karthichockalingam** [comment]
Hello @vladotomov 

I am looking into gslib/field-interp.cpp to implement the transfer functionality
I have to transfer a function from one mesh (source) to the second (target)

mesh_1 (or mesh_source or mesh_macro) and mesh_2  (or mesh_target or mesh_micro),  if you see the figure [Multiscale.pdf](https://github.com/user-attachments/files/16254245/Multiscale.pdf) the entire mesh_2 is smaller than one element of mesh_1.

The function I would like to transfer is the basis (or shape) functions of the mesh_1 for each element onto mesh_2.

My primary task would be to construct a source_func (i.e grid_function) for each shape function. So this grid function would be non-zero only one element of mesh_1. 

In short how best to loop over the elements in mesh_1 and generate a grid function for each of the basis functions?

I look forward to your response.

Kind regards,
Karthik.

**Comment by vladotomov** [comment]
Hi Karthik,

Here's one suggestion:
- You can make a list of quadrature points, on mesh_2, where the needed micro-computations will take place.
- Once you have the list, with their positions in physical space, you can generate the corresponding quadrature points on mesh_2.
- There, you can evaluate whatever macro basis functions are needed, and use the computed values to for the micro-calculations on mesh_2.

This avoids the need to generate complete `GridFunctions`, but only gets the needed values.

**Comment by karthichockalingam** [comment]
Thank you for getting back to me.

(Q1) In your first bullet point you meant to say mesh_1? 

I do like the idea of avoiding to generate ```GridFunctions``` for each macro (mesh_1) basis, but I don't need to evaluate the macro basis at the quadrature points of mesh_2 (micro) but rather at the boundaries of mesh_2 to impose it as Dirichlet condition. 

(Q2) Can I accomplish the above without creating ```GridFunctions```?

or

(Q3) Maybe in ```CalcShape(const IntegrationPoint &ip, Vector &shape) const``` I could pass the physical location of the boundary nodes of mesh_2 (micro) to evaluate the macro basis functions. I did find ```const Element *GetBdrElement(int i) const```  but not something which would retrieve the boundary nodes directly. 

(Q4) Just by getting the macro basis values at the boundary of mesh_2 (micro) - can I impose the Dirichlet condition without a  ```GridFunctions```?

**Comment by vladotomov** [comment]
(Q1): that's ok, you still need the physical locations of the boundary nodes of mesh_2 (for H1 spaces). For L2, you'd need to construct boundary integrals, which use quadrature points on the boundaries.

(Q2 and Q3): try with `FiniteElementSpace::GetBE(i)->GetNodes()`. This will give you the locations for H1 spaces.

(Q4): yes, you only need a value for every DOF. You'd need to do something like [this](https://github.com/mfem/mfem/blob/ce5b362077372840cd125ef334597ca5709c7976/examples/ex27.cpp#L240) for H1 (just assign the interpolated values to the corresponding DOFs).

(Q5): For a physical position on mesh_2, GSLIB will give you an element index and reference quadrature point on mesh_1, which can be used to evaluate.

**Comment by karthichockalingam** [comment]
I am almost there (just need some minor clarifications)

Yes, I am working with H1 Spaces. Here is the algorithm I plan to use.
 
**Step 1:** Locate all the physical nodal locations of the boundary nodes of mesh_2 (micro):
 
   ```FiniteElementSpace::GetBE(i)->GetNodes()```
 
   ```/// Get a const reference to the nodes of the element```
  ``` const IntegrationRule & GetNodes() const { return Nodes; }```
 
Since ``` GetNodes()``` returns type is ```IntegrationRule```: Does it return nodal location or integration point location? Since I am only after boundary nodal location.

On second thoughts, I probably need the locations of all nodes on mesh_2 and get it using ```vxyz = *mesh_2.GetNodes();``` so that I can reconstruct the ```gridfunction``` of the macro basis on mesh_2 (micro).

 
**Step 2:**  Evaluate the macro basis at the boundary nodes (or maybe at ALL nodes) of mesh_2 (micro).
 
Should I use ``` FiniteElement::CalcPhysShape(Trans, shape)``` or  ```Quad2DFiniteElement::CalcShape(ip, shape) const``` to evaluate the basis?
 
 
**Step 3:** Construct a grid function of the macro basis on mesh_2:
 
How to construct the grid function on mesh_2 of the evaluated macro basis from **step 2**? Is it using ``` GridFunction::SetFromTrueDofs```?
 
 
**Step 4:** Apply the Dirichlet boundary condition.
 
```u.ProjectBdrCoefficient(dbcCoef, dbc_bdr)```

 
Sorry, my final question is to why I need ``` For a physical position on mesh_2, GSLIB will give you an element index and reference quadrature point on mesh_1``` ?

Is it because in **Step 2** I can’t pass physical location but only ```reference quadrature point``` to evaluate the macro basis?

Big thank you!

**Comment by vladotomov** [comment]
Hi Karthik,

**Step 1:** `GetNodes()` would work, but it will return all DOF locations in physical space. If you need only the boundaries, then `GetBE(i)->GetNodes()` will give you a locations on the reference element, which can be combined with `ElementTransformation::Transform` to get the corresponding physical locations.

**Step 2:** These should be the same for H1. Just make sure to use the quadrature points given by `gslib`.

**Step 3:** You don't need to construct a full `GridFunction`, but just to set the values at the boundary DOFs.

**Step 4:** In example 27, `u.ProjectBdrCoefficient(dbcCoef, dbc_bdr)` simply sets values at the boundary DOFs. You'd need to replace this with the whole gslib procedure.

```My final question ... Is it because ...```
Yes. You need this to call `CalcShape` on mesh_1 in step 2.

**Comment by karthichockalingam** [comment]
Hi @vladotomov 

The code below works for the most part. But more cases have to be tested.

Q(1) While looping over the boundary edges of mesh_micro using ```mesh_micro.GetNBE()```,  then retrieve ```elem->GetNodes()```, I end up with the same nodes (that are shared between elements) multiple times. Is there a way I could avoid it? 

Q(2) How do I load the calculated ```sphape(0)``` (macro shape) values to the micro_mesh boundary dof (```dbcCoef```)so that I can set ```u.ProjectBdrCoefficient(dbcCoef, dbc_bdr)```?

Q(3) While using ```MakeCartesian2D``` is it possible to have the origin of the mesh at a point other than (0.0,0.0)?  I would like to move the micro mesh at the centre of one macro element. 

```
#include "mfem.hpp"
#include <fstream>

using namespace mfem;
using namespace std;

int main (int argc, char *argv[])
{

   Mesh mesh_macro = Mesh::MakeCartesian2D(2, 2, mfem::Element::Type::QUADRILATERAL,
                                     true, 2.0, 2.0);
   
   Mesh mesh_micro = Mesh::MakeCartesian2D(4, 4, mfem::Element::Type::QUADRILATERAL,
                                     true, 0.8, 0.8);
   
   int dim = mesh_macro.Dimension();

   mesh_micro.SetCurvature(1, false, dim, 0); 
   mesh_macro.SetCurvature(1, false, dim, 0); 

   FiniteElementCollection *fec = new H1_FECollection(1, dim);

   FiniteElementSpace fespace_macro(&mesh_macro, fec);
   FiniteElementSpace fespace_micro(&mesh_micro, fec);


   int bnd_pts = 2 * mesh_micro.GetNBE();
   Vector vxy(bnd_pts * dim);

   for (int i = 0; i < mesh_micro.GetNBE(); i++)
   {
      const FiniteElement * elem = fespace_micro.GetBE(i);
      const IntegrationRule & ir = elem->GetNodes();

      FaceElementTransformations * T = mesh_micro.GetBdrFaceTransformations(i);
      DenseMatrix P;
      T->Transform(ir, P);
      
      for (int j = 0; j < ir.Size(); j++)
      {
        //std::cout <<   P(0, j) << "  " << P(1, j) << std::endl;
        vxy((2*i)+j)           = P(0, j);
        vxy(bnd_pts + (2*i)+j) = P(1, j);
      }
   }
  
   FindPointsGSLIB finder_pos;
   finder_pos.FindPoints(mesh_macro, vxy);
   
   
   auto obj_elem = finder_pos.GetGSLIBElem();
   auto obj_ref_pos = finder_pos.GetReferencePosition();
   //auto obj_ref_pos = finder_pos.GetGSLIBReferencePosition();

   std::cout << "Boundary points:" << bnd_pts << std::endl;

   IntegrationPoint ip;
   BiLinear2DFiniteElement bilinear_elem;
   Vector shape(dim*dim);

   for (int i = 0; i < bnd_pts; i++)
   {
      ip.Set2(obj_ref_pos(i*dim + 0), obj_ref_pos(i*dim + 1));
      bilinear_elem.CalcShape(ip, shape);

      std::cout << obj_elem[i] << " " << obj_ref_pos(i*dim + 0) << "  " << obj_ref_pos(i*dim + 1) << "  " << shape(0) << std::endl;
   }
   
   return 0;
}
```

Many thanks!
Karthik.

**Comment by vladotomov** [comment]
Hi Karthik,

Q1: I can't think of anything, maybe not.

Q2: This is also related to Q1. You'll have to introduce some new data structures to keep track of what goes where, it's inevitable. I would create a marker array, based on you solution `GridFunction` DOF numbering, that tells which DOF is a boundary one, and which isn't. Then keep the coordinates of the boundary ones. And then get the shape values from the macro mesh and match them with that marker. This will allow you to set directly the boundary DOFs to the correct values.

Q3: See `Mesh::GetNodes()`. This gives you a `Vector` reference that you can manipulate to add displacements to the nodes.

**Comment by karthichockalingam** [comment]
Hi Vladimir,

Thank you for your response.

Sorry, I couldn't quite follow your suggestion ```based on your solution GridFunction DOF numbering```. (Q1) Did you mean to say you have to solve the problem first to get the solution?

Nevertheless,  I was able to accomplish what I wanted to using ``` fespace_micro.GetBdrElementDofs(i, dofs);```  (a functionality I was looking for and was unaware it existed previously) and new data structures.  I was able to generate the ```boundary DOFs``` and the corresponding ```shape values.```

```
 DOF # 0    1
 DOF # 1    0.8
 DOF # 2    0.6
 DOF # 3    0.4
 DOF # 4    0.2
 DOF # 21    0.16
 DOF # 20    0.2
 DOF # 22    0.12
 DOF # 23    0.08
 DOF # 24    0.04
 DOF # 5    0.8
 DOF # 10    0.6
 DOF # 15    0.4
 DOF # 9    0.16
 DOF # 14    0.12
 DOF # 19    0.08
```

(Q2) Can you please let me know what specific functionality to use and how to apply the Dirichlet boundary condition? 




Below is the entire code:

```
#include "mfem.hpp"
#include <fstream>

using namespace mfem;
using namespace std;

int main (int argc, char *argv[])
{

   Mesh mesh_macro = Mesh::MakeCartesian2D(2, 2, mfem::Element::Type::QUADRILATERAL,
                                     true, 2.0, 2.0);
   
   Mesh mesh_micro = Mesh::MakeCartesian2D(4, 4, mfem::Element::Type::QUADRILATERAL,
                                     true, 0.8, 0.8);
   
   int dim = mesh_macro.Dimension();

   mesh_micro.SetCurvature(1, false, dim, 0); 
   mesh_macro.SetCurvature(1, false, dim, 0); 

   FiniteElementCollection *fec = new H1_FECollection(1, dim);

   FiniteElementSpace fespace_macro(&mesh_macro, fec);
   FiniteElementSpace fespace_micro(&mesh_micro, fec);

   Array<int> dofs; //to store edge dofs

   set<int> dof_set, dof_set_cpy;

   for (int i = 0; i < mesh_micro.GetNBE(); i++)
   {
      const FiniteElement * elem = fespace_micro.GetBE(i);
      const IntegrationRule & ir = elem->GetNodes();
       
      fespace_micro.GetBdrElementDofs(i, dofs);

       for (int j = 0; j < dofs.Size(); j++)
      {
         dof_set.insert(dofs[j]);  
      }

   }

   //std::cout << "Bounday DOF Size " << dof_set.size() << std::endl;
   
   int bnd_pts =  dof_set.size();
   Vector vxy(bnd_pts * dim);
   int* dof_ids = new int[bnd_pts];
   int count = 0;

   for (int i = 0; i < dof_set.size(); i++)
   {
      const FiniteElement * elem = fespace_micro.GetBE(i);
      const IntegrationRule & ir = elem->GetNodes();
       
      fespace_micro.GetBdrElementDofs(i, dofs);

      FaceElementTransformations * T = mesh_micro.GetBdrFaceTransformations(i);
      DenseMatrix P;
      T->Transform(ir, P);

      assert(("The size of dof should be same ir", dofs.Size()  == ir.Size()));
      
      for (int j = 0; j < ir.Size(); j++)
      {
       // std::cout << "DOF# " << dofs[j] << "  "  << P(0, j) << "  " << P(1, j) << std::endl;
  
       if(dof_set_cpy.find(dofs[j]) == dof_set_cpy.end())
        {
        vxy(count)             = P(0, j);
        vxy(bnd_pts + (count)) = P(1, j);
        dof_set_cpy.insert(dofs[j]); 
        dof_ids[count] = dofs[j];
        count = count + 1;
        }
      }
   }
  
   FindPointsGSLIB finder_pos;
   finder_pos.FindPoints(mesh_macro, vxy);
   
   auto obj_elem = finder_pos.GetGSLIBElem();
   auto obj_ref_pos = finder_pos.GetReferencePosition();
   //auto obj_ref_pos = finder_pos.GetGSLIBReferencePosition();

   //std::cout << "Boundary points:" << bnd_pts << " Count = " << count << std::endl;

   IntegrationPoint ip;
   BiLinear2DFiniteElement bilinear_elem;
   Vector shape(dim*dim);

   for (int i = 0; i < bnd_pts; i++)
   {
      ip.Set2(obj_ref_pos(i*dim + 0), obj_ref_pos(i*dim + 1));
      bilinear_elem.CalcShape(ip, shape);

      //std::cout << obj_elem[i] << " DOF # " << dof_ids[i]  << "  " << obj_ref_pos(i*dim + 0) << "  " << obj_ref_pos(i*dim + 1) << "  " << shape(0) << std::endl;
      std::cout << " DOF # " << dof_ids[i]  << "    " << shape(0) << std::endl;
   }
   

   return 0;
}
```

**Comment by vladotomov** [comment]
Hi Karthik,

This code doesn't seem correct: you fill `dof_set` with DOF indices, but then you do
```
   for (int i = 0; i < dof_set.size(); i++)
   {
      const FiniteElement * elem = fespace_micro.GetBE(i);
```
which expects that `dof_set` is indexed over boundary elements. I don't see why you have two loops, all positions can be recorded in the first loop.

To your question, you need to have the following structure for the boundary dofs:
on micro: `dof_index_micro & dof_position_micro (x, y)`
find the corresponding value on macro to have: `dof_index_micro & dof_position_micro (x, y) & macro_value`
and then you set `solution_micro(dof_index_micro) = macro_value`

**Comment by karthichockalingam** [comment]
Hi Vladimir,

Thank you for pointing out the mistake I was making. I initially had two loops because, I didn't know the number of dofs on the boundary, which now get using ``` fespace_micro.GetBoundaryTrueDofs(boundary_dofs);```. I now have only one loop to record all the positions.

I get the desired output: 

```
 DOF # 0  x =  0  y =   0  macro_value =  1
 DOF # 1  x =  0.2  y =   0  macro_value =  0.8
 DOF # 2  x =  0.4  y =   0  macro_value =  0.6
 DOF # 3  x =  0.6  y =   0  macro_value =  0.4
 DOF # 4  x =  0.8  y =   0  macro_value =  0.2
 DOF # 21  x =  0.2  y =   0.8  macro_value =  0.16
 DOF # 20  x =  0  y =   0.8  macro_value =  0.2
 DOF # 22  x =  0.4  y =   0.8  macro_value =  0.12
 DOF # 23  x =  0.6  y =   0.8  macro_value =  0.08
 DOF # 24  x =  0.8  y =   0.8  macro_value =  0.04
 DOF # 5  x =  0  y =   0.2  macro_value =  0.8
 DOF # 10  x =  0  y =   0.4  macro_value =  0.6
 DOF # 15  x =  0  y =   0.6  macro_value =  0.4
 DOF # 9  x =  0.8  y =   0.2  macro_value =  0.16
 DOF # 14  x =  0.8  y =   0.4  macro_value =  0.12
 DOF # 19  x =  0.8  y =   0.6  macro_value =  0.08
```


(Q1) In you reponse you said ```find the corresponding value on macro to have: dof_index_micro & dof_position_micro (x, y) & macro_value``` but you don't use ```dof_position_micro (x, y)``` in ```solution_micro(dof_index_micro) = macro_value```?

(Q2) What is the data type of ```solution_micro``` and what **mfem functionality** do I pass ```solution_micro```?

Below is the code for your reference:

```
#include "mfem.hpp"
#include <fstream>

using namespace mfem;
using namespace std;

int main (int argc, char *argv[])
{

   Mesh mesh_macro = Mesh::MakeCartesian2D(2, 2, mfem::Element::Type::QUADRILATERAL,
                                     true, 2.0, 2.0);
   
   Mesh mesh_micro = Mesh::MakeCartesian2D(4, 4, mfem::Element::Type::QUADRILATERAL,
                                     true, 0.8, 0.8);
   
   int dim = mesh_macro.Dimension();

   mesh_micro.SetCurvature(1, false, dim, 0); 
   mesh_macro.SetCurvature(1, false, dim, 0); 

   FiniteElementCollection *fec = new H1_FECollection(1, dim);

   FiniteElementSpace fespace_macro(&mesh_macro, fec);
   FiniteElementSpace fespace_micro(&mesh_micro, fec);

   Array<int> boundary_dofs;
   fespace_micro.GetBoundaryTrueDofs(boundary_dofs);

   Array<int> dofs; //to store edge dofs

   set<int> dof_set;
   
   int bnd_pts =  boundary_dofs.Size();
   Vector vxy(bnd_pts * dim);
   int* dof_ids = new int[bnd_pts];
   int count = 0;

   for (int i = 0; i < mesh_micro.GetNBE(); i++)
   {
      const FiniteElement * elem = fespace_micro.GetBE(i);
      const IntegrationRule & ir = elem->GetNodes();
       
      fespace_micro.GetBdrElementDofs(i, dofs);

      FaceElementTransformations * T = mesh_micro.GetBdrFaceTransformations(i);
      DenseMatrix P;
      T->Transform(ir, P);

      assert(("The size of dof should be same ir", dofs.Size()  == ir.Size()));
      
      for (int j = 0; j < ir.Size(); j++)
      {
       // std::cout << "DOF# " << dofs[j] << "  "  << P(0, j) << "  " << P(1, j) << std::endl;
  
       if(dof_set.find(dofs[j]) == dof_set.end())
        {
        vxy(count)             = P(0, j);
        vxy(bnd_pts + (count)) = P(1, j);
        dof_set.insert(dofs[j]); 
        dof_ids[count] = dofs[j];
        count = count + 1;
        }
      }
   }
  
   FindPointsGSLIB finder_pos;
   finder_pos.FindPoints(mesh_macro, vxy);
   
   auto obj_elem = finder_pos.GetGSLIBElem();
   auto obj_ref_pos = finder_pos.GetReferencePosition();
   //auto obj_ref_pos = finder_pos.GetGSLIBReferencePosition();

   //std::cout << "Boundary points:" << bnd_pts << " Count = " << count << std::endl;

   IntegrationPoint ip;
   BiLinear2DFiniteElement bilinear_elem;
   Vector shape(dim*dim);

   for (int i = 0; i < bnd_pts; i++)
   {
      ip.Set2(obj_ref_pos(i*dim + 0), obj_ref_pos(i*dim + 1));
      bilinear_elem.CalcShape(ip, shape);

      //std::cout << obj_elem[i] << " DOF # " << dof_ids[i]  << "  " << obj_ref_pos(i*dim + 0) << "  " << obj_ref_pos(i*dim + 1) << "  " << shape(0) << std::endl;
      std::cout << " DOF # " << dof_ids[i]  << "  x =  " <<  vxy(i) << "  y =   " << vxy(bnd_pts + i) << "  macro_value =  " << shape(0) << std::endl;
   }


   return 0;
}
```

**Comment by vladotomov** [comment]
Q1: The `dof_position_micro (x, y)` is used to obtain the corresponding `macro_value`. It's not used in the final formula, but it's used to derive the needed value.

Q2: See how example 27 works. It's your solution `GridFunction`. As I wrote above, `Step 4: In example 27, u.ProjectBdrCoefficient(dbcCoef, dbc_bdr) simply sets values at the boundary DOFs. You'd need to replace this with the whole gslib procedure.`

Q3: Try with `GridFunction *vert_coord = mesh_micro.GetNodes();` and then edit the nodes through the pointer.

**Comment by karthichockalingam** [comment]
Thank you for your suggestions. I was able to solve a diffusion problem using the imposed macro basis function as boundary conditions. I will proceed next by spawning a micro mesh at every macro integration point.

**Comment by karthichockalingam** [comment]
Hi @vladotomov 

I noticed you closed the issue unfortunately, I haven't completed all the tasks yet. 

As I am spawning a micro-mesh at every macro integration point, my ```macro_mesh``` is of type ```ParMesh``` but ```FindPointsGSLIB::FindPoints``` would only take ```Mesh``` as argument type. 

Is there a way where I can pass the local portion of the  ```ParMesh``` as ``Mesh``` to ```FindPointsGSLIB::FindPoints```? 

Many thanks,
Karthik.
